Enable load_from_bytes of model and ExtScorer

bernardohenz opened the pull request:
This PR allows the loading of model and ExtScorer through array of bytes (```string```)

Some observations:

1. For this implementation to work, I had to use ```D_GLIBCXX_USE_CXX11_ABI=1``` when compiling the lib
2. I've added a new argument (```--init_from_bytes```) in the client for testing the loading
3. I do not know how you guys wanna handle the naming of methods, and if you will want to keep backwards compatibility (with ```DS_CreateModel``` for instance)
4. I only tested loading .pb models. I will need some help for implementing the loading of memmapped and tflite models.

lissyx commented on doc/BUILDING.rst:
```diff
@@ -73,7 +73,7 @@ You can now use Bazel to build the main DeepSpeech library, ``libdeepspeech.so``
 .. code-block::
 
-   bazel build --workspace_status_command="bash native_client/bazel_workspace_status_cmd.sh" --config=monolithic -c opt --copt=-O3 --copt="-D_GLIBCXX_USE_CXX11_ABI=0" --copt=-fvisibility=hidden //native_client:libdeepspeech.so
+   bazel build --workspace_status_command="bash native_client/bazel_workspace_status_cmd.sh" --config=monolithic -c opt --copt=-O3 --copt="-D_GLIBCXX_USE_CXX11_ABI=1" --copt=-fvisibility=hidden //native_client:libdeepspeech.so
```
This will cause compatibility issues on older distros

lissyx commented on native_client/Makefile:
```diff
@@ -19,7 +19,7 @@ clean:
 
 $(DEEPSPEECH_BIN): client.cc Makefile
-	$(CXX) $(CFLAGS) $(CFLAGS_DEEPSPEECH) $(SOX_CFLAGS) client.cc $(LDFLAGS) $(SOX_LDFLAGS)
+	$(CXX) $(CFLAGS) $(CFLAGS_DEEPSPEECH) $(SOX_CFLAGS) client.cc $(LDFLAGS) $(SOX_LDFLAGS) -llzma -lbz2
```
Why?

lissyx commented:
> For this implementation to work, I had to use `D_GLIBCXX_USE_CXX11_ABI=1` when compiling the lib

this will be problematic for compatibility with some systems

lissyx commented on native_client/tfmodelstate.cc:
```diff
@@ -59,13 +76,18 @@ TFModelState::init(const char* model_path)
-                             MemmappedFileSystem::kMemmappedPackageDefaultGraphDef,
-                             &graph_def_);
+  if (init_from_bytes){
+    // Need some help
```
Does it makes any sense at all?

lissyx commented on native_client/tflitemodelstate.cc:
```diff
@@ -156,23 +156,37 @@ getTfliteDelegates()
-    std::cerr << "Error at reading model file " << model_path << std::endl;
-    return DS_ERR_FAIL_INIT_MMAP;
+  if (init_from_bytes){
+    fbmodel_ = tflite::FlatBufferModel::BuildFromBuffer(model_string.c_str(), model_string.size_t);
```
How does that works, performance-wise?

lissyx replied:
When I see this, much alloc: https://github.com/tensorflow/tensorflow/blob/r2.3/tensorflow/lite/model_builder.cc#L104-L115

bernardohenz replied:
I haven't performed any benchmark, but I haven't noticed any slow down (for loading the pb model). When loading the Scorer, the test seems a bit slower due to reading/converting more than 900MB.
When running for our Language Model (size<20MB), it runs smoothly

lissyx replied:
Ok, if you use a 20MB scorer, yeah, it might not be too bad, but I guess it can also depend on use device you target. My fear is people using that without understanding the consequences and then complaining about memory usage.

lissyx commented:
@bernardohenz I fail to get the usecase where this can be useful, given how much this is going to force TensorFlow to allocate some memory.

reuben commented:
@lissyx #3152 is one example

lissyx commented:
> @lissyx #3152 is one example

Does it? From looking at the implem, I'm unsure this can work without trashing memory of the device: https://github.com/tensorflow/tensorflow/blob/r2.3/tensorflow/lite/model_builder.cc#L104-L115

lissyx commented:
> > I do not know how you guys wanna handle the naming of methods, and if you will want to keep backwards compatibility (with `DS_CreateModel` for instance)
> 
> Please introduce a new `DS_CreateModelFromBuffer` for example ?

Also, when you introduce that, please try and expose it in as much bindings as you can.

reuben commented:
> Does it? From looking at the implem, I'm unsure this can work without trashing memory of the device: https://github.com/tensorflow/tensorflow/blob/r2.3/tensorflow/lite/model_builder.cc#L104-L115

That's a good point, maybe we need to modify TensorFlow to take ownership of the pointer, or assume it'll outlive the session.

lissyx commented:
> > Does it? From looking at the implem, I'm unsure this can work without trashing memory of the device: https://github.com/tensorflow/tensorflow/blob/r2.3/tensorflow/lite/model_builder.cc#L104-L115
> 
> That's a good point, maybe we need to modify TensorFlow to take ownership of the pointer, or assume it'll outlive the session.

That's much more invasive and risky, except if we can upstream that, but we know how long it can take. Depending on @bernardohenz use-case, maybe it makes sense to just expose that for non protobuf mmap files ? or even only for TFLite models ?

lissyx commented:
> This PR allows the loading of model and ExtScorer through array of bytes (`string`)
> 
> Some observations:
> 
>     1. For this implementation to work, I had to use `D_GLIBCXX_USE_CXX11_ABI=1` when compiling the lib
> 
>     2. I've added a new argument (`--init_from_bytes`) in the client for testing the loading
> 
>     3. I do not know how you guys wanna handle the naming of methods, and if you will want to keep backwards compatibility (with `DS_CreateModel` for instance)
> 
>     4. I only tested loading .pb models. I will need some help for implementing the loading of memmapped and tflite models.

@bernardohenz I see you have code already for TFLite, can you describe what kind of help you need?

bernardohenz commented:
@lissyx I will be addressing the implementation of new methods for the API.
The points I should need some help is the implementation of loading from buffer of pdmm and tflite models. If I understood correctly, they would be much slower than loading from path, but still it's good to have this option than not having any option at all (we could put a message warning that this loading is slower for the sake of clarification).
Although I had implemented the code for tflite (just because I've found the ```tflite::FlatBufferModel::BuildFromBuffer```), I haven't tested it yet, I have never compiled with tflite and I will try that in the next week.

I have no answer for using ```D_GLIBCXX_USE_CXX11_ABI=1``` when compiling. When set to 0, the libs ended up having segmentation fault. I still haven't figured out what ```D_GLIBCXX_USE_CXX11_ABI``` does differently on what I've implemented. If you have any thoughts about this, it would be much appreciated.

lissyx commented:
> If I understood correctly, they would be much slower than loading from path, but still it's good to have this option than not having any option at all (we could put a message warning that this loading is slower for the sake of clarification).

I want to make sure we don't put ourselves in a position where people will misuse and have unrealistic expectations



> When set to 0, the libs ended up having segmentation fault. I still haven't figured out what `D_GLIBCXX_USE_CXX11_ABI` does differently on what I've implemented. If you have any thoughts about this, it would be much appreciated.

This does impact how C++ code handles strings, so maybe it depends on your env and in your case it makes sense? If you can share more of a stack, for example.

lissyx commented:
@bernardohenz can you:
 - check tflite code path on your side
 - force-push without `D_GLIBCXX_USE_CXX11_ABI`

I could take a look and run CI on that, so we can cross-check

lissyx commented:
> > If I understood correctly, they would be much slower than loading from path, but still it's good to have this option than not having any option at all (we could put a message warning that this loading is slower for the sake of clarification).
> 
> I want to make sure we don't put ourselves in a position where people will misuse and have unrealistic expectations

Maybe this should be warned about in the docs: "using this feature can have negative impact on the amount of memory consumed by your application"

lissyx commented:
@bernardohenz Your PR needs rebasing, it seems.

lissyx commented:
@bernardohenz Please don't change file's mode from 644 to 755.

lissyx commented on native_client/ctcdecode/scorer.cpp:
```diff
@@ -69,45 +71,60 @@ void Scorer::setup_char_map()
 }
 
-int Scorer::load_lm(const std::string& lm_path)
+int Scorer::load_lm(const std::string& lm_string, bool load_from_bytes)
```
Either provide a default value, split the implementation in two `load_lm` or update all call sites: this PR does not build because of `load_lm` calls in `generate_scorer_package.cpp`

bernardohenz replied:
I've provided a default value (false) for the boolean

lissyx commented on native_client/ctcdecode/scorer.cpp:
```diff
@@ -69,45 +71,60 @@ void Scorer::setup_char_map()
 }
 
-int Scorer::load_trie(std::ifstream& fin, const std::string& file_path)
+int Scorer::load_trie(std::stringstream& fin, const std::string& file_path, bool load_from_bytes)
```
`stringstream` is compatible with `ifstream` ?

bernardohenz replied:
As far as I tested yes

lissyx commented on native_client/deepspeech.cc:
```diff
@@ -265,6 +265,21 @@ StreamingState::processBatch(const vector<float>& buf, unsigned int n_steps)
+}
+
+int
+DS_CreateModel_(const std::string &aModelString,
```
Not sure how this is implemented, we go straight from `const char*` to `std::string &`

lissyx commented on native_client/deepspeech.h:
```diff
@@ -104,6 +106,33 @@ DEEPSPEECH_EXPORT
+ * @return Zero on success, non-zero on failure.
+ */
+DEEPSPEECH_EXPORT
+int DS_CreateModel_(const std::string &aModelString,
```
Nope, this should not be exposed.

lissyx commented on native_client/deepspeech.h:
```diff
@@ -104,6 +106,33 @@ DEEPSPEECH_EXPORT
+ * @return Zero on success, non-zero on failure.
+ */
+DEEPSPEECH_EXPORT
+int DS_CreateModelFromBuffer(const std::string &aModelBuffer,
```
`std::string` here makes that `C++` limited, the API is `C`

bernardohenz replied:
I'll be changing all the strings to ```char *```. But I'll have to pass the buffer size as parameter (I was having problems when a buffer encountered the ```\0``` in the middle of a buffer)

lissyx replied:
Thats perfectly fine.

lissyx commented on native_client/deepspeech.h:
```diff
@@ -157,6 +186,32 @@ DEEPSPEECH_EXPORT
+ * @return Zero on success, non-zero on failure (invalid arguments).
+ */
+DEEPSPEECH_EXPORT
+int DS_EnableExternalScorerFromBuffer(ModelState* aCtx,
```
Same, this is limiting to `C++` callers when we expose `C` API.

lissyx commented on native_client/deepspeech.h:
```diff
@@ -157,6 +186,32 @@ DEEPSPEECH_EXPORT
+ * @return Zero on success, non-zero on failure (invalid arguments).
+ */
+DEEPSPEECH_EXPORT                  
+int DS_EnableExternalScorer_(ModelState* aCtx,
```
This should not be exposed as well

bernardohenz replied:
Done

lissyx commented on native_client/kenlm/lm/bhiksha.cc:
```diff
@@ -17,9 +17,13 @@ DontBhiksha::DontBhiksha(const void * /*base*/, uint64_t /*max_offset*/, uint64_
 
 // TODO: put this in binary file header instead when I change the binary file format again.
-void ArrayBhiksha::UpdateConfigFromBinary(const BinaryFormat &file, uint64_t offset, Config &config) {
+void ArrayBhiksha::UpdateConfigFromBinary(const BinaryFormat &file, uint64_t offset, Config &config, bool load_from_bytes) {
```
Shouldn't this be upstreamed ?

lissyx commented:
@bernardohenz You should now be able to run PRs yourself.

lissyx commented:
@bernardohenz There are still failures related to `load_lm()` call sites on https://community-tc.services.mozilla.com/tasks/groups/VhPPLNsKSzeyZDg5kxjxqQ

lissyx commented:
> I've performed a small time benchmark for running the client (both for loading a .b and a .tflite):

Just for the sake of completeness, can you add figures for **before** your changes?



> I've fixed the API functions exposition, tested the loading for TFLite.

Good, so now it's just about:
 - exposing through all our bindings,
 - gettings builds green,
 - and adding some test coverage,

Right?

lissyx commented:
@bernardohenz I see some merges, can you make sure this branch is clean of merges ?

bernardohenz commented:
@lissyx yep, I'm trying to squash the commits, sorry.

reuben commented:
Just writing down some notes from my talk with Bernardo last week on the status of this PR:

- We fixed out a segfault due to passing an address of a temporary
- Bernardo is going to do add some timing calls in the scorer loading code to figure out if the slowdown is coming from the extra copy on the client.cc side (in which case it doesn't matter) or if the slowdown is inside libdeepspeech, in which case it needs to be addressed.
- Then we need to make sure all the bindings get covered and there's test coverage as mentioned by @lissyx above
